SchedulingManager failed. Exiting JVM when no nodes available

Hello,

we encounter an issue when network is not stable.
When our reaper loose connection to Cassandra DB, it fails and we have this issue :
`ERROR  [2020-11-09 23:22:08,476]  [SchedulingManagerTimer] i.c.s.SchedulingManager - SchedulingManager failed. Exiting JVM`
We are under last reaper version 2.1.2 at the writing time and on Ubuntu 16.04.
Our Cassandra is a single DC under version 3.11.3 with the following nodes :


```
10.234.167.14
10.234.167.15
10.234.167.18
10.234.167.19
10.234.167.20
10.234.167.21
```

Here the full output when the issue happens :


```
INFO   [2020-11-09 23:10:38,290]  [ReaperApplication-scheduler] i.c.ReaperApplication - Purged 0 repair runs from history
ERROR  [2020-11-09 23:22:08,470]  [SchedulingManagerTimer] i.c.s.SchedulingManager - failed managing schedule for run with id: null
ERROR  [2020-11-09 23:22:08,475]  [ReaperApplication-scheduler] i.c.ReaperApplication - Couldn't resume running repair runs
io.cassandrareaper.ReaperException: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /10.234.167.14:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.14:9042] Connection has been closed), /10.234.167.15:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.15:9042] Connection has been closed), /10.234.167.20:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.20:9042] Timed out waiting for server response), /10.234.167.18:9042 [only showing errors of first 3 hosts, use getErrors() for more details])
        at io.cassandrareaper.service.RepairManager.resumeRunningRepairRuns(RepairManager.java:141)
        at io.cassandrareaper.ReaperApplication.lambda$scheduleRepairManager$2(ReaperApplication.java:417)
        at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /10.234.167.14:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.14:9042] Connection has been closed), /10.234.167.15:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.15:9042] Connection has been closed), /10.234.167.20:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.20:9042] Timed out waiting for server response), /10.234.167.18:9042 [only showing errors of first 3 hosts, use getErrors() for more details])
        at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:83)
        at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:37)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:35)
        at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:293)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:58)
        at io.cassandrareaper.storage.CassandraStorage.getClusters(CassandraStorage.java:588)
        at io.cassandrareaper.storage.CassandraStorage.getRepairRunsWithState(CassandraStorage.java:930)
        at io.cassandrareaper.service.RepairManager.resumeRunningRepairRuns(RepairManager.java:134)
        ... 9 common frames omitted
Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /10.234.167.14:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.14:9042] Connection has been closed), /10.234.167.15:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.15:9042] Connection has been closed), /10.234.167.20:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.20:9042] Timed out waiting for server response), /10.234.167.18:9042 [only showing errors of first 3 hosts, use getErrors() for more details])
        at com.datastax.driver.core.RequestHandler.reportNoMoreHosts(RequestHandler.java:283)
        at com.datastax.driver.core.RequestHandler.access$1200(RequestHandler.java:61)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.findNextHostAndQuery(RequestHandler.java:375)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.retry(RequestHandler.java:563)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.processRetryDecision(RequestHandler.java:545)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onException(RequestHandler.java:947)
        at com.datastax.driver.core.Connection$Dispatcher.errorOutAllHandler(Connection.java:1368)
        at com.datastax.driver.core.Connection$ConnectionCloseFuture.force(Connection.java:1455)
        at com.datastax.driver.core.Connection$ConnectionCloseFuture.force(Connection.java:1436)
        at com.datastax.driver.core.Connection.defunct(Connection.java:619)
        at com.datastax.driver.core.Connection$12.fail(Connection.java:1432)
        at com.datastax.driver.core.Connection$12.onTimeout(Connection.java:1425)
        at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1633)
        at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
        at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
        at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        ... 1 common frames omitted
ERROR  [2020-11-09 23:22:08,476]  [SchedulingManagerTimer] i.c.s.SchedulingManager - catch exception
com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /10.234.167.14:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.14:9042] Timed out waiting for server response), /10.234.167.21:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.21:9042] Connection has been closed), /10.234.167.19:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.19:9042] Timed out waiting for server response))
        at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:83)
        at com.datastax.driver.core.exceptions.NoHostAvailableException.copy(NoHostAvailableException.java:37)
        at com.datastax.driver.core.DriverThrowables.propagateCause(DriverThrowables.java:35)
        at com.datastax.driver.core.DefaultResultSetFuture.getUninterruptibly(DefaultResultSetFuture.java:293)
        at com.datastax.driver.core.AbstractSession.execute(AbstractSession.java:58)
        at io.cassandrareaper.storage.CassandraStorage.getAllRepairSchedules(CassandraStorage.java:1370)
        at io.cassandrareaper.service.SchedulingManager.run(SchedulingManager.java:97)
        at java.util.TimerThread.mainLoop(Timer.java:555)
        at java.util.TimerThread.run(Timer.java:505)
Caused by: com.datastax.driver.core.exceptions.NoHostAvailableException: All host(s) tried for query failed (tried: /10.234.167.14:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.14:9042] Timed out waiting for server response), /10.234.167.21:9042 (com.datastax.driver.core.exceptions.TransportException: [/10.234.167.21:9042] Connection has been closed), /10.234.167.19:9042 (com.datastax.driver.core.exceptions.OperationTimedOutException: [/10.234.167.19:9042] Timed out waiting for server response))
        at com.datastax.driver.core.RequestHandler.reportNoMoreHosts(RequestHandler.java:283)
        at com.datastax.driver.core.RequestHandler.access$1200(RequestHandler.java:61)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.findNextHostAndQuery(RequestHandler.java:375)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.retry(RequestHandler.java:563)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.processRetryDecision(RequestHandler.java:545)
        at com.datastax.driver.core.RequestHandler$SpeculativeExecution.onException(RequestHandler.java:947)
        at com.datastax.driver.core.Connection$Dispatcher.errorOutAllHandler(Connection.java:1368)
        at com.datastax.driver.core.Connection$ConnectionCloseFuture.force(Connection.java:1455)
        at com.datastax.driver.core.Connection$ConnectionCloseFuture.force(Connection.java:1436)
        at com.datastax.driver.core.Connection.defunct(Connection.java:619)
        at com.datastax.driver.core.Connection$12.fail(Connection.java:1432)
        at com.datastax.driver.core.Connection$12.onTimeout(Connection.java:1425)
        at com.datastax.driver.core.Connection$ResponseHandler$1.run(Connection.java:1633)
        at io.netty.util.HashedWheelTimer$HashedWheelTimeout.expire(HashedWheelTimer.java:663)
        at io.netty.util.HashedWheelTimer$HashedWheelBucket.expireTimeouts(HashedWheelTimer.java:738)
        at io.netty.util.HashedWheelTimer$Worker.run(HashedWheelTimer.java:466)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
ERROR  [2020-11-09 23:22:08,476]  [SchedulingManagerTimer] i.c.s.SchedulingManager - SchedulingManager failed. Exiting JVM.
```



┆Issue is synchronized with this [Jira Task](https://k8ssandra.atlassian.net/browse/K8SSAND-258) by [Unito](https://www.unito.io)
┆Issue Number: K8SSAND-258

--- a/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
+++ b/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
@@ -126,7 +126,7 @@ public final class SchedulingManager extends TimerTask {
         LOG.error("Error while scheduling repairs due to a connection problem with the database", e);
       } catch (Throwable ex) {
         if (lastId == null) {
-          LOG.error("Failed to managing repair schedules", ex);
+          LOG.error("Failed managing repair schedules", ex);
         } else {
           LOG.error("Failed to managing repair schedule with id '{}'", lastId, ex);
         }

--- a/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
+++ b/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
@@ -128,7 +128,7 @@ public final class SchedulingManager extends TimerTask {
         if (lastId == null) {
           LOG.error("Failed managing repair schedules", ex);
         } else {
-          LOG.error("Failed to managing repair schedule with id '{}'", lastId, ex);
+          LOG.error("Failed managing repair schedule with id '{}'", lastId, ex);
         }
         try {
           assert false : "if assertions are enabled then exit the jvm";

--- a/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
+++ b/src/server/src/main/java/io/cassandrareaper/service/SchedulingManager.java
@@ -34,6 +34,7 @@ import java.util.TimerTask;
 import java.util.UUID;
 import java.util.concurrent.ThreadLocalRandom;
 
+import com.datastax.driver.core.exceptions.DriverException;
 import com.datastax.driver.core.exceptions.DriverInternalError;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.Supplier;
@@ -121,9 +122,14 @@ public final class SchedulingManager extends TimerTask {
         }
       } catch (DriverInternalError expected) {
         LOG.debug("Driver connection closed, Reaper is shutting down.");
+      } catch (DriverException e) {
+        LOG.error("Error while scheduling repairs due to a connection problem with the database", e);
       } catch (Throwable ex) {
-        LOG.error("failed managing schedule for run with id: {}", lastId);
-        LOG.error("catch exception", ex);
+        if (lastId == null) {
+          LOG.error("Failed to managing repair schedules", ex);
+        } else {
+          LOG.error("Failed to managing repair schedule with id '{}'", lastId, ex);
+        }
         try {
           assert false : "if assertions are enabled then exit the jvm";
         } catch (AssertionError ae) {

